core/src/main/java/io/questdb/cutlass/text/TextImportTask.java

```diff
@@ -0,0 +1,987 @@
+                throws TextException {
+            TypeAdapter type = this.types.getQuick(fieldIndex);
+            try {
+                type.write(w, fieldIndex, dbcs, utf8Sink);
+            } catch (Exception ignore) {
+                errors++;
+                logError(offset, fieldIndex, dbcs);
+                switch (atomicity) {
+                    case Atomicity.SKIP_ALL:
+                        tableWriterRef.rollback();
+                        throw TextException.$("bad syntax [line offset=").put(offset).put(",column=").put(fieldIndex).put(']');
+                    case Atomicity.SKIP_ROW:
+                        w.cancel();
+                        return true;
+                    default: // SKIP column
+                        break;
+                }
+            }
+            return false;
+        }
+
+        private void onFieldsPartitioned(final ObjList<DirectByteCharSequence> values, int valuesLength, DirectCharSink utf8Sink) {
+            assert tableWriterRef != null;
+            DirectByteCharSequence dbcs = values.getQuick(timestampIndex);
+            final TableWriter.Row w = getRow(dbcs, offset);
+            if (w == null) {
+                return;
+            }
+            for (int i = 0; i < valuesLength; i++) {
+                dbcs = values.getQuick(i);
+                if (i == timestampIndex || dbcs.length() == 0) {
+                    continue;
+                }
+                if (onField(offset, dbcs, w, i, utf8Sink)) return;
+            }
+            w.append();
+        }
+
+        private TableWriter.Row getRow(DirectByteCharSequence dbcs, long offset) {
+            try {
+                return tableWriterRef.newRow(timestampAdapter.getTimestamp(dbcs));
+            } catch (Exception e) {
+                if (atomicity == Atomicity.SKIP_ALL) {
+                    throw TextException.$("Can't parse timestamp on line starting at offset=").put(offset).put(". ").put(e.getMessage());
+                } else {
+                    logError(offset, timestampIndex, dbcs);
+                    return null;
+                }
+            }
+        }
+
+        private long openIndexChunks(FilesFacade ff, Path partitionPath, DirectLongList mergeIndexes, int partitionLen) {
+            long mergedIndexSize = 0;
+            long chunk = ff.findFirst(partitionPath);
+            if (chunk > 0) {
+                try {
+                    do {
+                        // chunk loop
+                        long chunkName = ff.findName(chunk);
+                        long chunkType = ff.findType(chunk);
+                        if (chunkType == Files.DT_FILE) {
+                            partitionPath.trimTo(partitionLen);
+                            partitionPath.concat(chunkName).$();
+
+                            final long fd = TableUtils.openRO(ff, partitionPath, LOG);
```

Comment:
Possible mmaped regions leak due to TableUtils.openRO exception

Reply:
Fixed in latest commit . There's one additional failure mode code handles (direct list reallocation fails while adding mmap address or size) now but it's difficult to trigger before Alex's PR with mem limit is merged .